MTV-1722: AIO buffer performance (#598)
kubev2v/forklift-documentation: documentation/modules/mtv-aio-buffer.adoc

// Module included in the following assemblies:
//
// * documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc

:_content-type: CONCEPT
[id="mtv-aio-buffer_{context}"]
= Increasing asynchronous I/O (AIO) sizes and buffer counts for NBD transport mode

This document describes how to change NBD transport NFC parameters for increased migration performance when using the {project-first} product.

[WARNING]
.Cold Migration Only
====

> Using AIO buffering is only suitable for Cold Migration use cases.

Reviewer: We saw  that if AIO is configured and we try warm warm will fail...
Should we also add something ike "Desbale AOI settings before initilazing Warm Migration - Refer  chapter 9.11.3. "Disabling AIO Buffer Configuration"

[id="mtv-aio-performance-results_{context}"]
== Performance Testing Results

There were several extensive tests, which were conducted across multiple configurations:

* Single VMware ESXi server compared to multiple ESXi servers.
* Single virtual machine (VM) compared to multiple VMs.

[id="mtv-aio-buffer-key-findings_{context}"]
=== Key findings

* The best migration performance was achieved by migrating using multiple VMs (10) on a single ESXi host with the following values:
** `VixDiskLib.nfcAio.Session.BufSizeIn64KB=16`
** `vixDiskLib.nfcAio.Session.BufCount=4`

* The following improvements were noted when using AIO buffer (Asynchronous Buffer Counts) settings:
** Migration time was reduced by *31.1%*, from `0:24:32` to `0:16:54`.
** Read rate was increased from `347.83MB/s` to `504.93MB/s`.

* There was no significant improvement observed when using AIO buffer settings with a single VM.

* There was no significant improvement observed when using AIO buffer settings with multiple VMs from multiple hosts.

[id="mtv-enable-aio-buffer_{context}"]

> == Enabling AIO buffer configuration

Reviewer 2: need to edit section 9.11.2
since the forklift controller name prefix is dynamic , and on the current example its hard coded
so please change it to this :

`Validating Controller Pod support for AIO values

Ensure that the forklift-controller pod in the openshift-mtv namespace supports the AIO buffer values.
Since the pod name prefix is dynamic , check the pod name first by running the folowing command:

oc get pods -n openshift-mtv | grep forklift-controller | awk '{print $1}'
output : forklift-controller-667f57c8f8-qllnx
This is the pod name prefix from the example : forklift-controller-667f57c8f8-qllnx

Check the environment variables of the pod by running:
`oc get pod forklift-controller-667f57c8f8-qllnx -n openshift-mtv -o yaml``

.Validating Controller Pod support for AIO values

> * Ensure that the `forklift-controller` pod in the `openshift-mtv` namespace supports the AIO buffer values. Check the environment variables of the pod by running:

Author: ```suggestion
* Ensure that the `forklift-controller` pod in the `openshift-mtv` namespace supports the AIO buffer values.
+
Since the pod name prefix is dynamic, check the pod name first by running the following command:
```

> 

Author: ```suggestion

+
[source,terminal]
----
oc get pods -n openshift-mtv | grep forklift-controller | awk '{print $1}'
----

+
[source,terminal]
----
oc get pod forklift-controller-5b45f66949-vgzhx -n openshift-mtv -o yaml
----
+

* Check for the following lines in the output:
+
[source,termina]
----
...
\- name: VIRT\_V2V\_EXTRA\_ARGS
\- name: VIRT\_V2V\_EXTRA\_CONF\_CONFIG\_MAP
...
----

.Editing ForkliftController Configuration

* In the `openshift-mtv namespace`, edit the `ForkliftController` object to include the AIO buffer values.
+
Add the following under the spec section:
+
[source,termina]
----
oc edit forkliftcontroller -n openshift-mtv`
----

* For example, to add the following values:
** virt_v2v_extra_args:
+
[source,termina]
----
--vddk-config /mnt/extra-v2v-conf/input.conf
----
+
[source,termina]
----

> virt_v2v_extra_conf_config_map: "`perf`"

Reviewer 2: @anarnold97 
please change the following line with this one : 
virt_v2v_extra_conf_config_map: "perf"

// Module included in the following assemblies:
//
// * documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc

:_content-type: CONCEPT
[id="mtv-aio-buffer_{context}"]
= Increasing asynchronous I/O (AIO) sizes and buffer counts for NBD transport mode

This document describes how to change NBD transport NFC parameters for increased migration performance when using the {project-first} product.

> 

Author: ```suggestion

[WARNING]
.Cold Migration Only
====
Using AIO buffering is only suitable for Cold Migration use cases. 
====

```

// Module included in the following assemblies:
//
// * documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc

:_content-type: CONCEPT
[id="mtv-aio-buffer_{context}"]
= AIO buffer counts for customer usage

This document is intended to summarize the performance team's testing on the impact of changes to the AIO buffer count, specifically to assess any effects on datastore read rates and overall migration time when using the {project-first} product.

.Additional information

* link:https://docs.vmware.com/en/VMware-vSphere/7.0/vsphere-vddk-programming-guide/GUID-5D166ED1-7205-4110-8D72-0C51BB63CC3D.html[Best Practices for NBD Transport]
* link:https://forums.veeam.com/vmware-vsphere-f24/solution-for-poor-nbd-performance-t93084.html[Solution for poor NBD performance]

[id="mtv-aio-performance-results_{context}"]
== Performance Testing Results

There were several extensive tests, which were conducted across multiple configurations:

* Single VMware ESXi server compared to multiple ESXi servers.
* Single virtual machine (VM) compared to multiple VMs.

[id="mtv-aio-buffer-key-findings_{context}"]
=== Key findings

> The best migration performance was achieved by migrating using Multiple 10 VMs(10) on a single ESXi host with the following values:

Reviewer: Multiple VMs (10)

Reviewer: no need 10 before VMs

* `VixDiskLib.nfcAio.Session.BufSizeIn64KB=16`
* `vixDiskLib.nfcAio.Session.BufCount=4`

> The following improvements were noted when using AIO buffer (Asynchronous Buffer Counts) settings:

Reviewer: I suggest  indent right as it relates to the above Many VMs on single host

Author: @dgur - changes made and preview manually updated (Netlify has ended the license). thanks

Reviewer: looks good now.
I need now that Tzahi and Mordechai validate the commands to set and remove it

* Migration time was reduced by *31.1%*, from `0:24:32` to `0:16:54`.
* Read rate was increased from `347.83MB/s` to `504.93MB/s`.

> There was no significant improvement observed when using AIO buffer settings with a single VM.

Reviewer: Can you somehow bullet it and the below?

Since the pod name prefix is dynamic, check the pod name first by running the following command:
+
[source,terminal]
----
oc get pods -n openshift-mtv | grep forklift-controller | awk '{print $1}'
----
+
The example output is as follows:
+
[source,termina]
----
forklift-controller-667f57c8f8-qllnx
----
+
[NOTE]
====
This is the pod name prefix from the example: `forklift-controller-667f57c8f8-qllnx`
====

* Check the environment variables of the pod by running:
+
[source,termina]
----
oc get pod forklift-controller-667f57c8f8-qllnx -n openshift-mtv -o yaml
----

> ----

Author: ```suggestion
```

* There was no significant improvement observed when using AIO buffer settings with multiple VMs from multiple hosts.

[id="mtv-enable-aio-buffer_{context}"]
== Enabling AIO buffer configuration

.Validating Controller Pod support for AIO values

* Ensure that the `forklift-controller` pod in the `openshift-mtv` namespace supports the AIO buffer values.
+
Since the pod name prefix is dynamic, check the pod name first by running the following command:
+
[source,terminal]
----
oc get pods -n openshift-mtv | grep forklift-controller | awk '{print $1}'
----
+
The example output is as follows:
+
[source,termina]
----
forklift-controller-667f57c8f8-qllnx
----
+
[NOTE]
====
This is the pod name prefix from the example: `forklift-controller-667f57c8f8-qllnx`
====

* Check the environment variables of the pod by running:
+
[source,termina]
----
oc get pod forklift-controller-667f57c8f8-qllnx -n openshift-mtv -o yaml
----
+
* Check for the following lines in the output:
+
[source,termina]
----
...
\- name: VIRT\_V2V\_EXTRA\_ARGS
\- name: VIRT\_V2V\_EXTRA\_CONF\_CONFIG\_MAP
...
----

.Editing ForkliftController Configuration

> * In the `openshift-mtv namespace`, edit the `ForkliftController` object to include the AIO buffer values.

Author: ```suggestion
* In the `openshift-mtv namespace`, edit the `ForkliftController` object to include the AIO buffer values, by running the following command:
++
[source,termina]
----
oc edit forkliftcontroller -n openshift-mtv
----
```

> +

Reviewer 2: add below line 84 , the command to edit the ForkliftController as  follow : 
run the following command to edit the ForkliftController object :
oc edit forkliftcontroller -n openshift-mtv

* Ensure that the `forklift-controller` pod in the `openshift-mtv` namespace supports the AIO buffer values.
+
Since the pod name prefix is dynamic, check the pod name first by running the following command:
+
[source,terminal]
----
oc get pods -n openshift-mtv | grep forklift-controller | awk '{print $1}'
----
+
The example output is as follows:
+
[source,termina]
----
forklift-controller-667f57c8f8-qllnx
----
+
[NOTE]
====
This is the pod name prefix from the example: `forklift-controller-667f57c8f8-qllnx`
====

* Check the environment variables of the pod by running:
+
[source,termina]
----
oc get pod forklift-controller-667f57c8f8-qllnx -n openshift-mtv -o yaml
----
+
[source,terminal]
----
oc get pod forklift-controller-5b45f66949-vgzhx -n openshift-mtv -o yaml
----
+
* Check for the following lines in the output:
+
[source,termina]
----
...
\- name: VIRT\_V2V\_EXTRA\_ARGS
\- name: VIRT\_V2V\_EXTRA\_CONF\_CONFIG\_MAP
...
----

.Editing ForkliftController Configuration

* In the `openshift-mtv namespace`, edit the `ForkliftController` object to include the AIO buffer values.
+
Add the following under the spec section:
+
[source,termina]
----

> oc edit forkliftcontroller -n openshift-mtv`

Author: ```suggestion
virt_v2v_extra_args: --vddk-config /mnt/extra-v2v-conf/input.conf
virt_v2v_extra_conf_config_map: "perf"
```

// Module included in the following assemblies:
//
// * documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc

:_content-type: CONCEPT
[id="mtv-aio-buffer_{context}"]
= AIO buffer counts for customer usage

The purpose of this document is to summarize the performance team's testing on the impact of changes to the AIO buffer (Asynchronous Buffer Counts) specifically to assess any effects on datastore read rates and overall migration time when using project-first.

.Additional information

* link:https://docs.vmware.com/en/VMware-vSphere/7.0/vsphere-vddk-programming-guide/GUID-5D166ED1-7205-4110-8D72-0C51BB63CC3D.html[Best Practices for NBD Transport]
* link:https://forums.veeam.com/vmware-vsphere-f24/solution-for-poor-nbd-performance-t93084.html[Solution for poor NBD performance]

[id="mtv-aio-performance-results_{context}"]
== Performance Testing Results

* Single VMware ESXi server compared to multiple ESXi servers.
* Single virtual machine (VM) compared to multiple VMs.

[id="mtv-aio-buffer-key-findings_{context}"]
=== Key Findings

The highest migration performance was achieved with 10 VMs on a single ESXi host with the following values:

> 
> * `VixDiskLib.nfcAio.Session.BufSizeIn64KB=16`
> * `vixDiskLib.nfcAio.Session.BufCount=4`
> 
> 
> [id="mtv-aio-buffer-improvement_{context}"]
> === Improvement:

Author: ```suggestion
[id="mtv-aio-buffer-improvement_{context}"]
== Improvement:
```

.Validating Controller Pod support for AIO values

* Ensure that the `forklift-controller` pod in the `openshift-mtv` namespace supports the AIO buffer values.
+
Since the pod name prefix is dynamic, check the pod name first by running the following command:
+
[source,terminal]
----
oc get pods -n openshift-mtv | grep forklift-controller | awk '{print $1}'
----
+
The example output is as follows:
+
[source,termina]
----
forklift-controller-667f57c8f8-qllnx
----
+
[NOTE]
====
This is the pod name prefix from the example: `forklift-controller-667f57c8f8-qllnx`
====

* Check the environment variables of the pod by running:
+
[source,termina]
----
oc get pod forklift-controller-667f57c8f8-qllnx -n openshift-mtv -o yaml
----

> +
> [source,terminal]
> ----
> oc get pod forklift-controller-5b45f66949-vgzhx -n openshift-mtv -o yaml

Author: ```suggestion
```

// Module included in the following assemblies:
//
// * documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc

:_content-type: CONCEPT
[id="mtv-aio-buffer_{context}"]
= AIO buffer counts for customer usage

The purpose of this document is to summarize the performance team's testing on the impact of changes to the AIO buffer (Asynchronous Buffer Counts) specifically to assess any effects on datastore read rates and overall migration time when using project-first.

.Additional information

* link:https://docs.vmware.com/en/VMware-vSphere/7.0/vsphere-vddk-programming-guide/GUID-5D166ED1-7205-4110-8D72-0C51BB63CC3D.html[Best Practices for NBD Transport]
* link:https://forums.veeam.com/vmware-vsphere-f24/solution-for-poor-nbd-performance-t93084.html[Solution for poor NBD performance]

[id="mtv-aio-performance-results_{context}"]
== Performance Testing Results

> There were several extensive tests, which were conducted across multiple configurations:

Reviewer: I don think we need to specify how we tested exactly  - it is our problem

[id="mtv-aio-buffer-key-findings_{context}"]
=== Key findings

* The best migration performance was achieved by migrating using multiple VMs (10) on a single ESXi host with the following values:
** `VixDiskLib.nfcAio.Session.BufSizeIn64KB=16`
** `vixDiskLib.nfcAio.Session.BufCount=4`

* The following improvements were noted when using AIO buffer (Asynchronous Buffer Counts) settings:
** Migration time was reduced by *31.1%*, from `0:24:32` to `0:16:54`.
** Read rate was increased from `347.83MB/s` to `504.93MB/s`.

* There was no significant improvement observed when using AIO buffer settings with a single VM.

* There was no significant improvement observed when using AIO buffer settings with multiple VMs from multiple hosts.

[id="mtv-enable-aio-buffer_{context}"]
== Enabling AIO buffer configuration

.Validating Controller Pod support for AIO values

* Ensure that the `forklift-controller` pod in the `openshift-mtv` namespace supports the AIO buffer values.
+
Since the pod name prefix is dynamic, check the pod name first by running the following command:

+
[source,terminal]
----
oc get pods -n openshift-mtv | grep forklift-controller | awk '{print $1}'
----

> +

Author: ```suggestion
+
The example output is as follows:
+
[source,termina]
----
forklift-controller-667f57c8f8-qllnx
----
+
[NOTE]
====
This is the pod name prefix from the example: `forklift-controller-667f57c8f8-qllnx`
====

* Check the environment variables of the pod by running:
+
[source,termina]
----
oc get pod forklift-controller-667f57c8f8-qllnx -n openshift-mtv -o yaml
----
+
```

----
+
The example output is as follows:
+
[source,termina]
----
VixDiskLib.nfcAio.Session.BufSizeIn64KB=16
vixDiskLib.nfcAio.Session.BufCount=4
----

.Enabling Debugging (optional)

* To enable debug logs, convert the configuration to Base64, including a high log level:
+
[source,termina]
----
echo -e
"`VixDiskLib.nfcAio.Session.BufSizeIn64KB=16\nVixDiskLib.nfcAio.Session.BufCount=4\nVixDiskLib.nfc.LogLevel=4`"
| base64
----
+
[NOTE]
====
Adding a high log level will degrade performance and is for debugging purposes only.
====

[id="mtv-disable-aio-buffer_{context}"]
== Disabling AIO Buffer Configuration

To disable the AIO buffer configuration, complete the following steps:

* Edit the ForkliftController Object: Remove the previously added lines from the spec section in the ForkliftController object:
+
[source,termina]
----
oc edit forkliftcontroller -n openshift-mtv
----

* Remove the following lines:
+
[source,termina]
----
virt_v2v_extra_args: "`–vddk-config /mnt/extra-v2v-conf/input.conf`"
virt_v2v_extra_conf_config_map: "`perf`"
----
. Delete the ConfigMap: Remove the perf ConfigMap that was created
earlier:

`oc delete cm perf -n openshift-mtv`

....

>   3\.	Restart the Forklift Controller Pod (Optional):

Reviewer 2: from this line please remove the "3\."
  3\.	Restart the Forklift Controller Pod (Optional):

need to be 
Restart the Forklift Controller Pod (Optional):

....

If needed, ensure the changes take effect by restarting the forklift-controller pod.

[id="mtv-aio-buffer-key-requirements_{context}"]
=== Key requirements for AIO Buffer (Asynchronous Buffer Counts) support

.Key requirements for AIO Buffer support
[width="100%",cols="54%,46%",options="header",]
|===
|Version |Support

|vSphere Version
a|
vSphere 6.7 Update 3 and later.
Fully supported in vSphere 7.x versions, including 7.0.x.

> |VDDK Version

Reviewer 2: this section missing  the next recommendation for VDDK : 

VDDK Version:
Requires VDDK 6.7.3 or newer. The VDDK version typically aligns with the ESXi version installed on the vSphere host.

Reviewer 2: me and daniel made some changes to this section please use the update one : 

VDDK /vSpheres Versions: 

Tested with : 
vSphere : 7.0.3
VDDK : 7.0.3 
* for other  VDDK / vSpheres versions please check the AIO buffer support in the official VMware documentation

[id="mtv-enable-aio-buffer_{context}"]
== Enabling AIO buffer configuration

.Validating Controller Pod support for AIO values

* Ensure that the `forklift-controller` pod in the `openshift-mtv` namespace supports the AIO buffer values.
+
Since the pod name prefix is dynamic, check the pod name first by running the following command:
+
[source,terminal]
----
oc get pods -n openshift-mtv | grep forklift-controller | awk '{print $1}'
----
+
The example output is as follows:
+
[source,termina]
----
forklift-controller-667f57c8f8-qllnx
----
+
[NOTE]
====
This is the pod name prefix from the example: `forklift-controller-667f57c8f8-qllnx`
====

* Check the environment variables of the pod by running:
+
[source,termina]
----
oc get pod forklift-controller-667f57c8f8-qllnx -n openshift-mtv -o yaml
----
+
* Check for the following lines in the output:
+
[source,termina]
----
...
\- name: VIRT\_V2V\_EXTRA\_ARGS
\- name: VIRT\_V2V\_EXTRA\_CONF\_CONFIG\_MAP
...
----

.Editing ForkliftController Configuration

* In the `openshift-mtv namespace`, edit the `ForkliftController` object to include the AIO buffer values, by running the following command:

> ++

Author: ```suggestion
+
```

[id="mtv-enable-aio-buffer_{context}"]
== Enabling AIO buffer configuration

.Validating Controller Pod support for AIO values

* Ensure that the `forklift-controller` pod in the `openshift-mtv` namespace supports the AIO buffer values.
+
Since the pod name prefix is dynamic, check the pod name first by running the following command:
+
[source,terminal]
----
oc get pods -n openshift-mtv | grep forklift-controller | awk '{print $1}'
----
+
The example output is as follows:
+
[source,termina]
----
forklift-controller-667f57c8f8-qllnx
----
+
[NOTE]
====
This is the pod name prefix from the example: `forklift-controller-667f57c8f8-qllnx`
====

* Check the environment variables of the pod by running:
+
[source,termina]
----
oc get pod forklift-controller-667f57c8f8-qllnx -n openshift-mtv -o yaml
----
+
* Check for the following lines in the output:
+
[source,termina]
----
...
\- name: VIRT\_V2V\_EXTRA\_ARGS
\- name: VIRT\_V2V\_EXTRA\_CONF\_CONFIG\_MAP
...
----

.Editing ForkliftController Configuration

* In the `openshift-mtv namespace`, edit the `ForkliftController` object to include the AIO buffer values.
+
Add the following under the spec section:
+
[source,termina]
----

> virt_v2v_extra_args: --vddk-config /mnt/extra-v2v-conf/input.conf

Reviewer 2: change lines 88+89 
with those  new  lines content 
virt_v2v_extra_args: "--vddk-config /mnt/extra-v2v-conf/input.conf"
virt_v2v_extra_conf_config_map: "perf"

virt_v2v_extra_conf_config_map: "perf"
----

.Creating a ConfigMap mamed perf

* Create the required ConfigMap using the following command:
+
[source,termina]
----

> oc -n openshift-mtv create cm perf`

Reviewer 2: replace this line: oc -n openshift-mtv create cm perf`
with this line: oc -n openshift-mtv create cm perf

----

.Preparing the ConfigMap content

* Convert the desired buffer configuration values to Base64. For example, for 16/4:
+
[source,termina]
----

> echo -e "`VixDiskLib.nfcAio.Session.BufSizeIn64KB=16\nvixDiskLib.nfcAio.Session.BufCount=4`" | base64`

Reviewer 2: replace this old line: 
 echo -e "`VixDiskLib.nfcAio.Session.BufSizeIn64KB=16\nvixDiskLib.nfcAio.Session.BufCount=4`" | base64`
 
with this  line:
echo -e "VixDiskLib.nfcAio.Session.BufSizeIn64KB=16\nvixDiskLib.nfcAio.Session.BufCount=4" | base64

.Verifying migration logs

* Run a migration plan and check the logs of the migration pod. Confirm that the AIO buffer settings are passed as parameters, particularly the `--vddk-config value`.
+
For example:
+
[source,termina]
----
exec: /usr/bin/virt-v2v … --vddk-config /mnt/extra-v2v-conf/input.conf
----

+
Sample log excerpt:
+
[source,termina]
----
Buffer size calc for 16 value:
(16 * 64 * 1024 = 1048576)
nbdkit: vddk[1]: debug: [NFC VERBOSE] NfcAio_OpenSession:
Opening an AIO session.
nbdkit: vddk[1]: debug: [NFC INFO] NfcAioInitSession:
Disabling
read-ahead buffer since the AIO buffer size of 1048576 is >=
the read-ahead buffer size of 65536. Explicitly setting flag
'`NFC_AIO_SESSION_NO_NET_READ_AHEAD`'
nbdkit: vddk[1]: debug: [NFC VERBOSE] NfcAioInitSession: AIO Buffer Size is 1048576
nbdkit: vddk[1]: debug: [NFC VERBOSE] NfcAioInitSession: AIO Buffer
Count is 4
----
+
[NOTE]
====
The above logs were when using `debug_level = 4`
====

.Inspecting ConfigMap values Content are in the Migration Pod

* Log in to the migration pod and verify the buffer settings using the following command:
+
[source,termina]
----

> `cat /mnt/extra-v2v-conf/input.conf`

Reviewer 2: replace  this line:  `cat /mnt/extra-v2v-conf/input.conf`
with this one:  cat /mnt/extra-v2v-conf/input.conf

+
[source,termina]
----
echo -e
"`VixDiskLib.nfcAio.Session.BufSizeIn64KB=16\nVixDiskLib.nfcAio.Session.BufCount=4\nVixDiskLib.nfc.LogLevel=4`"
| base64
----
+
[NOTE]
====
Adding a high log level will degrade performance and is for debugging purposes only.
====

[id="mtv-disable-aio-buffer_{context}"]
== Disabling AIO Buffer Configuration

To disable the AIO buffer configuration, complete the following steps:

* Edit the ForkliftController Object: Remove the previously added lines from the spec section in the ForkliftController object:
+
[source,termina]
----
oc edit forkliftcontroller -n openshift-mtv
----

* Remove the following lines:
+
[source,termina]
----
virt_v2v_extra_args: "`–vddk-config /mnt/extra-v2v-conf/input.conf`"
virt_v2v_extra_conf_config_map: "`perf`"
----
. Delete the ConfigMap: Remove the perf ConfigMap that was created
earlier:

[source,termina]
----
oc delete cm perf -n openshift-mtv
----

.	Restart the Forklift Controller Pod (Optional).

If needed, ensure the changes take effect by restarting the forklift-controller pod.

[id="mtv-aio-buffer-key-requirements_{context}"]
== Key requirements for AIO Buffer (Asynchronous Buffer Counts) support

> .Key requirements for AIO Buffer support

Reviewer 2: replace this title: 9.11.4. Key requirements for AIO Buffer (Asynchronous Buffer Counts) support 

with 9.11.4  VDDK /vSpheres Versions:

> [width="100%",cols="54%,46%",options="header",]

Reviewer 2: replace lines 243-261 
with this new content : 

Tested with : 
       - vSphere : 7.0.3
       - VDDK : 7.0.3 
         * for other  VDDK / vSpheres versions please check the AIO buffer support in the official VMware 
            documentation

+
[source,terminal]
----
oc get pods -n openshift-mtv | grep forklift-controller | awk '{print $1}'
----
+
The example output is as follows:
+
[source,termina]
----
forklift-controller-667f57c8f8-qllnx
----
+
[NOTE]
====
This is the pod name prefix from the example: `forklift-controller-667f57c8f8-qllnx`
====

* Check the environment variables of the pod by running:
+
[source,termina]
----
oc get pod forklift-controller-667f57c8f8-qllnx -n openshift-mtv -o yaml
----
+
[source,terminal]
----
oc get pod forklift-controller-5b45f66949-vgzhx -n openshift-mtv -o yaml
----
+
* Check for the following lines in the output:
+
[source,termina]
----
...
\- name: VIRT\_V2V\_EXTRA\_ARGS
\- name: VIRT\_V2V\_EXTRA\_CONF\_CONFIG\_MAP
...
----

.Editing ForkliftController Configuration

* In the `openshift-mtv namespace`, edit the `ForkliftController` object to include the AIO buffer values.
+
Add the following under the spec section:
+
[source,termina]
----
oc edit forkliftcontroller -n openshift-mtv`
----

* For example, to add the following values:
** virt_v2v_extra_args:
+
[source,termina]
----
--vddk-config /mnt/extra-v2v-conf/input.conf
----
+
[source,termina]
----

> virt_v2v_extra_conf_config_map: "`perf`"

Author: ```suggestion
virt_v2v_extra_conf_config_map: "perf"
```

* The following improvements were noted when using AIO buffer (Asynchronous Buffer Counts) settings:
** Migration time was reduced by *31.1%*, from `0:24:32` to `0:16:54`.
** Read rate was increased from `347.83MB/s` to `504.93MB/s`.

* There was no significant improvement observed when using AIO buffer settings with a single VM.

* There was no significant improvement observed when using AIO buffer settings with multiple VMs from multiple hosts.

[id="mtv-enable-aio-buffer_{context}"]
== Enabling AIO buffer configuration

.Validating Controller Pod support for AIO values

* Ensure that the `forklift-controller` pod in the `openshift-mtv` namespace supports the AIO buffer values.
+
Since the pod name prefix is dynamic, check the pod name first by running the following command:
+
[source,terminal]
----
oc get pods -n openshift-mtv | grep forklift-controller | awk '{print $1}'
----
+
The example output is as follows:
+
[source,termina]
----
forklift-controller-667f57c8f8-qllnx
----
+
[NOTE]
====
This is the pod name prefix from the example: `forklift-controller-667f57c8f8-qllnx`
====

* Check the environment variables of the pod by running:
+
[source,termina]
----

> oc get pod forklift-controller-667f57c8f8-qllnx -n openshift-mtv -o yaml

Reviewer 2: remove this line please  , its using the wrong pod prefix name

Author: @TzahiAshkenazi - Please can you confirm the line number to be removed

Reviewer 2: this comment was related to previous  line you already removed with the second pod prefix name 
so according to the new PREVIEW no need to remove

----
+
[source,terminal]
----
oc get pod forklift-controller-5b45f66949-vgzhx -n openshift-mtv -o yaml
----
+
* Check for the following lines in the output:
+
[source,termina]
----
...
\- name: VIRT\_V2V\_EXTRA\_ARGS
\- name: VIRT\_V2V\_EXTRA\_CONF\_CONFIG\_MAP
...
----

.Editing ForkliftController Configuration

* In the `openshift-mtv namespace`, edit the `ForkliftController` object to include the AIO buffer values.
+
Add the following under the spec section:
+
[source,termina]
----
oc edit forkliftcontroller -n openshift-mtv`
----

> * For example, to add the following values:
> ** virt_v2v_extra_args:
> +
> [source,termina]
> ----
> --vddk-config /mnt/extra-v2v-conf/input.conf
> ----
> +
> [source,termina]
> ----
> virt_v2v_extra_conf_config_map: "perf"
> ----

Author: ```suggestion

```

// Module included in the following assemblies:
//
// * documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc

:_content-type: CONCEPT
[id="mtv-aio-buffer_{context}"]

> = AIO buffer counts for customer usage

Author: ```suggestion
= Increasing asynchronous I/O (AIO) sizes and buffer counts for NBD transport mode
```
Sorry, but that title is too long, so i will abbreviate

> = AIO buffer counts for customer usage

Author: Apologies but this title is a bit too long and could create problems when building the docs

So as a compromise could we abbreviate to **Increasing asynchronous I/O (AIO) sizes and buffer counts for NBD transport mode**

> This document is intended to summarize the performance team's testing on the impact of changes to the AIO buffer count, specifically to assess any effects on datastore read rates and overall migration time when using the {project-first} product.

Reviewer: The gdoc was to summarize testing - in the official docs we just need to explain what to do and why - I believe.
Let's write something like this

This chapter describes how to change NBD transport  NFC parameters for increased migration performance.

> This document is intended to summarize the performance team's testing on the impact of changes to the AIO buffer count, specifically to assess any effects on datastore read rates and overall migration time when using the {project-first} product.

Author: ```suggestion
This document describes how to change NBD transport NFC parameters for increased migration performance when using the {project-first} product.
```

.Additional information

* link:https://docs.vmware.com/en/VMware-vSphere/7.0/vsphere-vddk-programming-guide/GUID-5D166ED1-7205-4110-8D72-0C51BB63CC3D.html[Best Practices for NBD Transport]
* link:https://forums.veeam.com/vmware-vsphere-f24/solution-for-poor-nbd-performance-t93084.html[Solution for poor NBD performance]

[id="mtv-aio-performance-results_{context}"]
== Performance Testing Results

> There were several extensive tests, which were conducted across multiple configurations:

Reviewer: Remove. We are not doing test report

> * Single VMware ESXi server compared to multiple ESXi servers.

Reviewer: remove

> * Single virtual machine (VM) compared to multiple VMs.

Reviewer: remove

+
[source,terminal]
----
oc get pods -n openshift-mtv | grep forklift-controller | awk '{print $1}'
----
+
The example output is as follows:
+
[source,termina]
----
forklift-controller-667f57c8f8-qllnx
----
+
[NOTE]
====
This is the pod name prefix from the example: `forklift-controller-667f57c8f8-qllnx`
====

* Check the environment variables of the pod by running:
+
[source,termina]
----
oc get pod forklift-controller-667f57c8f8-qllnx -n openshift-mtv -o yaml
----
+
[source,terminal]
----
oc get pod forklift-controller-5b45f66949-vgzhx -n openshift-mtv -o yaml
----
+
* Check for the following lines in the output:
+
[source,termina]
----
...
\- name: VIRT\_V2V\_EXTRA\_ARGS
\- name: VIRT\_V2V\_EXTRA\_CONF\_CONFIG\_MAP
...
----

.Editing ForkliftController Configuration

* In the `openshift-mtv namespace`, edit the `ForkliftController` object to include the AIO buffer values.
+
Add the following under the spec section:
+
[source,termina]
----
oc edit forkliftcontroller -n openshift-mtv`
----

* For example, to add the following values:
** virt_v2v_extra_args:
+
[source,termina]
----
--vddk-config /mnt/extra-v2v-conf/input.conf
----
+
[source,termina]
----

> virt_v2v_extra_conf_config_map: "`perf`"

Author: ```suggestion
virt_v2v_extra_conf_config_map: "perf"
```

Disabling
read-ahead buffer since the AIO buffer size of 1048576 is >=
the read-ahead buffer size of 65536. Explicitly setting flag
'`NFC_AIO_SESSION_NO_NET_READ_AHEAD`'
nbdkit: vddk[1]: debug: [NFC VERBOSE] NfcAioInitSession: AIO Buffer Size is 1048576
nbdkit: vddk[1]: debug: [NFC VERBOSE] NfcAioInitSession: AIO Buffer
Count is 4
----
+
[NOTE]
====
The above logs were when using `debug_level = 4`
====

.Inspecting ConfigMap values Content are in the Migration Pod

* Log in to the migration pod and verify the buffer settings using the following command:
+
[source,termina]
----
`cat /mnt/extra-v2v-conf/input.conf`
----
+
The example output is as follows:
+
[source,termina]
----
VixDiskLib.nfcAio.Session.BufSizeIn64KB=16
vixDiskLib.nfcAio.Session.BufCount=4
----

.Enabling Debugging (optional)

* To enable debug logs, convert the configuration to Base64, including a high log level:
+
[source,termina]
----
echo -e
"`VixDiskLib.nfcAio.Session.BufSizeIn64KB=16\nVixDiskLib.nfcAio.Session.BufCount=4\nVixDiskLib.nfc.LogLevel=4`"
| base64
----
+
[NOTE]
====
Adding a high log level will degrade performance and is for debugging purposes only.
====

> .Disabling AIO Buffer Configuration

Reviewer: Disable Part shouldn’t  be under Unable but another paragraph

vixDiskLib.nfcAio.Session.BufCount=4
----

.Enabling Debugging (optional)

* To enable debug logs, convert the configuration to Base64, including a high log level:
+
[source,termina]
----
echo -e
"`VixDiskLib.nfcAio.Session.BufSizeIn64KB=16\nVixDiskLib.nfcAio.Session.BufCount=4\nVixDiskLib.nfc.LogLevel=4`"
| base64
----
+
[NOTE]
====
Adding a high log level will degrade performance and is for debugging purposes only.
====

.Disabling AIO Buffer Configuration

To disable the AIO buffer configuration, complete the following steps:

* Edit the ForkliftController Object: Remove the previously added lines from the spec section in the ForkliftController object:
+
[source,termina]
----
oc edit forkliftcontroller -n openshift-mtv
----

* Remove the following lines:
+
[source,termina]
----
virt_v2v_extra_args: "`–vddk-config /mnt/extra-v2v-conf/input.conf`"
virt_v2v_extra_conf_config_map: "`perf`"
----
. Delete the ConfigMap: Remove the perf ConfigMap that was created
earlier:

`oc delete cm perf -n openshift-mtv`

....
  3\.	Restart the Forklift Controller Pod (Optional):
....

If needed, ensure the changes take effect by restarting the forklift-controller pod.

> [id="mtv-aio-buffer-comparison_{context}"]
> == AIO buffer comparisons
> 
> [id="mtv-aio-buffer-comparison-single-vm_{context}"]
> === One VM Migration with AIO (16/4) and without AIO
> 
> .One VM Migration with AIO and without AIO
> [width="99%",cols="20%,12%,20%,16%,16%,16%",options="header",]
> |===
> |AIO Buffer values |MTV |VDDK |max_vm inflight |Migration type |Total
> duration
> |N/A
> |2.7.4
> |7.0.3
> |20
> |cold
> |00:08:12

Author: ```suggestion
```

.Additional information

* link:https://docs.vmware.com/en/VMware-vSphere/7.0/vsphere-vddk-programming-guide/GUID-5D166ED1-7205-4110-8D72-0C51BB63CC3D.html[Best Practices for NBD Transport]
* link:https://forums.veeam.com/vmware-vsphere-f24/solution-for-poor-nbd-performance-t93084.html[Solution for poor NBD performance]

[id="mtv-aio-performance-results_{context}"]
== Performance Testing Results

Several extensive tests were conducted across multiple configurations.

* Single ESXi server versus multiple ESXi servers.
* Single VM versus multiple VMs.

[id="mtv-aio-buffer-key-findings_{context}"]
=== Key Findings

* The best migration performance was with 10 VMs on a single ESXi.
host with the following values:
** VixDiskLib.nfcAio.Session.BufSizeIn64KB=16
** vixDiskLib.nfcAio.Session.BufCount=4

[id="mtv-aio-buffer-improvement_{context}"]
=== Improvement:

* Migration time was reduced by *31.1%*, from `0:24:32` to
`0:16:54`.
* Read rate increased from `347.83MB/s` to `504.93MB/s`.

[id="mtv-aio-buffer-observations_{context}"]
== Observations

* No significant impact was observed when using AIO buffer settings with a single VM.

> [id="mtv-aio-buffer-environment_{context}"]
> === Environment
> 
> * The RDU2 scale lab utilizes metal OpenShift clusters and VMware vSphere servers, which employ dedicated hardware, isolated networking, and storage to eliminate external variables and guarantee consistent performance metrics.
> ** *Vmware information*:
> *** vSphere Version: 7.0.3
> *** ESXi Version: 7.0.3
> *** DataStore : ISCSI & FC Luns, attached to netapp storage
> ** *MTV information*:
> *** VDDK Version: 7.0.3
> *** MTV Version: 2.7.4
> *** Max VM Inflight: 20
> ** *VM Specification*:
> *** *Network*: vSwitch with MTU 9000, 10 GbE ESXi interfaces.
> *** *Disks*: Single disk, 50GB size, 70% utilization.
> 

Author: ```suggestion
```
